### PR TITLE
BUILD-281: add library-go watch/update of new OCM feature gate field

### PR DESCRIPTION
**Pull request by gabemontero**
build csi volumes (to facilitate builds with RH entitlements) and the underlying csi driver shared resource CSO operator are tech preview feature gate items for the upcoming release

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -3,22 +3,25 @@ module github.com/openshift/cluster-openshift-controller-manager-operator
...
+	k8s.io/api v0.22.1
+	k8s.io/apimachinery v0.22.1
+	k8s.io/client-go v0.22.1
+	k8s.io/component-base v0.22.1
 	k8s.io/klog/v2 v2.9.0
 )
 
 replace vbom.ml/util => github.com/fvbommel/util v0.0.0-20180919145318-efcd4e0f9787
+
+replace github.com/openshift/api => github.com/gabemontero/api v0.0.0-20210910154316-eaa29e1f16df
```
Comment (gabemontero) [inline go.mod:25]: this captures the latest state of openshift/api#1007

when that merges, we'll just change the version for openshift/api up above

File: go.mod [hunk 2]
```diff
@@ -3,22 +3,25 @@ module github.com/openshift/cluster-openshift-controller-manager-operator
...
+
+replace github.com/openshift/library-go => github.com/gabemontero/library-go v0.0.0-20210910154638-8565ba189275
```
Comment (gabemontero) [inline go.mod:27]: this captures the latest state of https://github.com/openshift/library-go/pull/1201

when that merges, we'll just change the version for openshift/library-go up above

File: pkg/operator/configobservation/configobservercontroller/observe_config_controller.go [hunk 3]
```diff
@@ -48,6 +50,9 @@ func NewConfigObserver(
 		builds.ObserveBuildControllerConfig,
 		network.ObserveExternalIPAutoAssignCIDRs,
 		deployimages.ObserveControllerManagerImagesConfig,
+		//TODO when https://github.com/openshift/api/pull/982 merges we can add "CSIDriverSharedResource" to that
+		// currently includes "BuildCSIVolumes"
```
Comment (gabemontero) [inline pkg/operator/configobservation/configobservercontroller/observe_config_controller.go:54]: @adambkaplan - conceivably, we don't need both CSIDriverSharedResource that you cited in BUILD-292 and BuildCSIVolumes that you cited in BUILD-281, unless you think CSI shared resource stuff could move out of tech preview mode in a release prior to when Build's consumption of it move out of tech preview mode.

**Comment by gabemontero** [comment]
the e2e-aws-operator failure needs investigation .... not immediately apparent to me what caused that

**Comment by gabemontero** [comment]
Probably related to 

`pods/openshift-controller-manager-operator_openshift-controller-manager-operator-8b94fc78-h8x6z_openshift-controller-manager-operator_previous.log.gz:E0910 17:03:36.813554       1 runtime.go:78] Observed a panic: &runtime.TypeAssertionError{_interface:(*runtime._type)(nil), concrete:(*runtime._type)(0x24befe0), asserted:(*runtime._type)(0x22f9cc0), missingMethod:"FeatureGateLister"} (interface conversion: configobservation.Listers is not featuregates.FeatureGateLister: missing method FeatureGateLister)`

**Comment by gabemontero** [comment]
Full stack:

```
2021-09-10T17:03:57.359070088Z E0910 17:03:57.359041       1 runtime.go:78] Observed a panic: &runtime.TypeAssertionError{_interface:(*runtime._type)(nil), concrete:(*runtime._type)(0x24befe0), asserted:(*runtime._type)(0x22f9cc0), missingMethod:"FeatureGateLister"} (interface conversion: configobservation.Listers is not featuregates.FeatureGateLister: missing method FeatureGateLister)
2021-09-10T17:03:57.359070088Z goroutine 702 [running]:
2021-09-10T17:03:57.359070088Z k8s.io/apimachinery/pkg/util/runtime.logPanic(0x22f2c60, 0xc0015a4270)
2021-09-10T17:03:57.359070088Z 	k8s.io/apimachinery@v0.22.1/pkg/util/runtime/runtime.go:74 +0x95
2021-09-10T17:03:57.359070088Z k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0xc00100bd58, 0x1, 0x1)
2021-09-10T17:03:57.359070088Z 	k8s.io/apimachinery@v0.22.1/pkg/util/runtime/runtime.go:48 +0x86
2021-09-10T17:03:57.359070088Z panic(0x22f2c60, 0xc0015a4270)
2021-09-10T17:03:57.359070088Z 	runtime/panic.go:965 +0x1b9
2021-09-10T17:03:57.359070088Z github.com/openshift/library-go/pkg/operator/configobserver/featuregates.(*featureFlags).ObserveFeatureFlags(0xc0009828a0, 0x2987ad8, 0xc000582ba0, 0x29d4638, 0xc000856240, 0xc0015a4090, 0x0, 0x0, 0x0, 0x0)
2021-09-10T17:03:57.359070088Z 	github.com/openshift/library-go@v0.0.0-20210817120645-b59564e63303/pkg/operator/configobserver/featuregates/observe_featuregates.go:49 +0xcb
2021-09-10T17:03:57.359070088Z github.com/openshift/library-go/pkg/operator/configobserver.ConfigObserver.sync(0xc0009828d0, 0x6, 0x6, 0x29c88f8, 0xc000993a60, 0x2987ad8, 0xc000582ba0, 0x0, 0x0, 0x0, ...)
2021-09-10T17:03:57.359070088Z 	github.com/openshift/library-go@v0.0.0-20210817120645-b59564e63303/pkg/operator/configobserver/config_observer_controller.go:139 +0x370
2021-09-10T17:03:57.359070088Z github.com/openshift/library-go/pkg/controller/factory.(*baseController).reconcile(0xc000b80900, 0x29bf970, 0xc000eb5c80, 0x29ba810, 0xc0015a4000, 0x0, 0x0)
2021-09-10T17:03:57.359070088Z 	github.com/openshift/library-go@v0.0.0-20210817120645-b59564e63303/pkg/controller/factory/base_controller.go:205 +0x5c
2021-09-10T17:03:57.359070088Z github.com/openshift/library-go/pkg/controller/factory.(*baseController).processNextWorkItem(0xc000b80900, 0x29bf970, 0xc000eb5c80)
2021-09-10T17:03:57.359070088Z 	github.com/openshift/library-go@v0.0.0-20210817120645-b59564e63303/pkg/controller/factory/base_controller.go:264 +0x253
2021-09-10T17:03:57.359070088Z github.com/openshift/library-go/pkg/controller/factory.(*baseController).runWorker.func1(0x29bf970, 0xc000eb5c80)
2021-09-10T17:03:57.359070088Z 	github.com/openshift/library-go@v0.0.0-20210817120645-b59564e63303/pkg/controller/factory/base_controller.go:196 +0xb9
2021-09-10T17:03:57.359070088Z k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext.func1()
2021-09-10T17:03:57.359070088Z 	k8s.io/apimachinery@v0.22.1/pkg/util/wait/wait.go:185 +0x37
2021-09-10T17:03:57.359070088Z k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0xc0012abef8)
2021-09-10T17:03:57.359070088Z 	k8s.io/apimachinery@v0.22.1/pkg/util/wait/wait.go:155 +0x5f
2021-09-10T17:03:57.359070088Z k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00100bef8, 0x296ba80, 0xc000156a50, 0xc000eb5c01, 0xc000ec1f80)
2021-09-10T17:03:57.359070088Z 	k8s.io/apimachinery@v0.22.1/pkg/util/wait/wait.go:156 +0x9b
2021-09-10T17:03:57.359070088Z k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000c74ef8, 0x3b9aca00, 0x0, 0x1, 0xc000ec1f80)
2021-09-10T17:03:57.359070088Z 	k8s.io/apimachinery@v0.22.1/pkg/util/wait/wait.go:133 +0x98
2021-09-10T17:03:57.359070088Z k8s.io/apimachinery/pkg/util/wait.JitterUntilWithContext(0x29bf970, 0xc000eb5c80, 0xc000c74f58, 0x3b9aca00, 0x0, 0x1)
2021-09-10T17:03:57.359070088Z 	k8s.io/apimachinery@v0.22.1/pkg/util/wait/wait.go:185 +0xa6
2021-09-10T17:03:57.359070088Z k8s.io/apimachinery/pkg/util/wait.UntilWithContext(...)
2021-09-10T17:03:57.359070088Z 	k8s.io/apimachinery@v0.22.1/pkg/util/wait/wait.go:99
2021-09-10T17:03:57.359070088Z github.com/openshift/library-go/pkg/controller/factory.(*baseController).runWorker(0xc000b80900, 0x29bf970, 0xc000eb5c80)
2021-09-10T17:03:57.359070088Z 	github.com/openshift/library-go@v0.0.0-20210817120645-b59564e63303/pkg/controller/factory/base_controller.go:187 +0x74
2021-09-10T17:03:57.359070088Z github.com/openshift/library-go/pkg/controller/factory.(*baseController).Run.func2(0xc000b80900, 0xc0012837b0, 0x29bf970, 0xc000eb5c80)
2021-09-10T17:03:57.359070088Z 	github.com/openshift/library-go@v0.0.0-20210817120645-b59564e63303/pkg/controller/factory/base_controller.go:117 +0x75
2021-09-10T17:03:57.359070088Z created by github.com/openshift/library-go/pkg/controller/factory.(*baseController).Run
2021-09-10T17:03:57.359070088Z 	github.com/openshift/library-go@v0.0.0-20210817120645-b59564e63303/pkg/controller/factory/base_controller.go:112 +0x2f6
2021-09-10T17:03:57.359143072Z I0910 17:03:57.359117       1 event.go:282] Event(v1.ObjectReference{Kind:"Deployment", Namespace:"openshift-controller-manager-operator", Name:"openshift-controller-manager-operator", UID:"8fc9ca17-7b15-4ce3-9a69-6922853ec228", APIVersion:"apps/v1", ResourceVersion:"", FieldPath:""}): type: 'Warning' reason: 'Openshift-Controller-Manager-OperatorPanic' Panic observed: interface conversion: configobservation.Listers is not featuregates.FeatureGateLister: missing method FeatureGateLister
```

**Comment by gabemontero** [comment]
now down to unrelated sig flakes

/test e2e-aws

File: pkg/operator/configobservation/configobservercontroller/observe_config_controller.go (hunk 3 above)
Comment (gabemontero) [reply pkg/operator/configobservation/configobservercontroller/observe_config_controller.go:55]: UPDATE: @adambkaplan and I talked and he reminded me that long term we will want to support multiple CSI volume types for builds

hence the multiple feature gates do make sense.

**Comment by gabemontero** [comment]
/assign @adambkaplan 

OK I've vendored in the merged openshift/api and openshift/library-go changes for our new openshiftcontrolplane feature gate []string field

**Comment by gabemontero** [comment]
e2e-aws perm fail seems unrelated ... let's see

/test e2e-aws

File: pkg/operator/configobservation/interfaces.go [hunk 4]
```diff
@@ -13,6 +13,7 @@ type Listers struct {
 	BuildConfigLister  configlistersv1.BuildLister
 	ConfigMapLister    corelistersv1.ConfigMapLister
 	NetworkLister      configlistersv1.NetworkLister
+	FeatureGateLister_ configlistersv1.FeatureGateLister
```
Comment (adambkaplan) [inline pkg/operator/configobservation/interfaces.go:16]: Accidental underscore?
```suggestion
	FeatureGateLister  configlistersv1.FeatureGateLister
```

File: pkg/operator/configobservation/interfaces.go [hunk 5]
```diff
@@ -23,3 +24,7 @@ func (l Listers) ResourceSyncer() resourcesynccontroller.ResourceSyncer {
 func (l Listers) PreRunHasSynced() []cache.InformerSynced {
 	return l.PreRunCachesSynced
 }
+
+func (l Listers) FeatureGateLister() configlistersv1.FeatureGateLister {
+	return l.FeatureGateLister_
```
Comment (adambkaplan) [inline pkg/operator/configobservation/interfaces.go:29]:
```suggestion
	return l.FeatureGateLister
```

File: pkg/operator/configobservation/configobservercontroller/observe_config_controller.go [hunk 6]
```diff
@@ -48,6 +51,7 @@ func NewConfigObserver(
 		builds.ObserveBuildControllerConfig,
 		network.ObserveExternalIPAutoAssignCIDRs,
 		deployimages.ObserveControllerManagerImagesConfig,
+		featuregates.NewObserveFeatureFlagsFunc(sets.NewString("BuildCSIVolumes"), sets.String{}, []string{"featureGates"}),
```
Comment (adambkaplan) [inline pkg/operator/configobservation/configobservercontroller/observe_config_controller.go:54]: looks good

File: pkg/operator/configobservation/configobservercontroller/observe_config_controller.go [hunk 7]
```diff
@@ -30,10 +32,11 @@ func NewConfigObserver(
...
 		eventRecorder,
 		configobservation.Listers{
-			ImageConfigLister: configInformers.Config().V1().Images().Lister(),
-			BuildConfigLister: configInformers.Config().V1().Builds().Lister(),
-			NetworkLister:     configInformers.Config().V1().Networks().Lister(),
-			ConfigMapLister:   kubeInformersForOperatorNamespace.Core().V1().ConfigMaps().Lister(),
+			ImageConfigLister:  configInformers.Config().V1().Images().Lister(),
+			BuildConfigLister:  configInformers.Config().V1().Builds().Lister(),
+			NetworkLister:      configInformers.Config().V1().Networks().Lister(),
+			FeatureGateLister_: configInformers.Config().V1().FeatureGates().Lister(),
```
Comment (adambkaplan) [inline pkg/operator/configobservation/configobservercontroller/observe_config_controller.go:38]:
```suggestion
			FeatureGateLister: configInformers.Config().V1().FeatureGates().Lister(),
```

**Review by adambkaplan — APPROVED** [review]
/approve

Nits on the underscore, but otherwise looks good.

**Comment by adambkaplan** [comment]
/assign @jkhelil 

Assigning follow up review to Jawed since he looked into the build controller enhancement before sprint planning.

**Comment by gabemontero** [comment]
> /approve
> 
> Nits on the underscore, but otherwise looks good.

underscores are actually need @adambkaplan or you get a compile error

after getting stuck with this, I saw the analogous fix in the kube api server controller

**Comment by gabemontero** [comment]
> > /approve
> > Nits on the underscore, but otherwise looks good.
> 
> underscores are actually needed @adambkaplan or you get a compile error
> 
> after getting stuck with this, I saw the analogous fix in the kube api server controller
> 
> I'll try and (re)find the link

yeah here is the link:  https://github.com/openshift/cluster-kube-apiserver-operator/blob/master/pkg/operator/configobservation/interfaces.go#L17-L24

File: go.mod (hunk 2 above)
Comment (jkhelil) [reply go.mod]: This is merged now @gabemontero you can update it

File: go.mod (hunk 1 above)
Comment (jkhelil) [reply go.mod]: This is merged now @gabemontero you can update it

**Pull request merged**